src/databases.cpp

```diff
@@ -289,42 +333,240 @@ const int16_t *PairDistanceKVectorDatabase::FindPairsExact(const Catalog &catalo
-           && catalog[pairs[liberalLowerIndex*2]].spatial * catalog[pairs[liberalLowerIndex*2+1]].spatial >= maxQueryCos
-        )
-    { liberalLowerIndex++; }
+    // There's no good reason to be using >= and <= for the comparison against max/min, but the
+    // tests fail otherwise (because they use angle, with its acos, instead of forward cos like us).
+    // It's an insignificant difference.
+    while (liberalLowerIndex < liberalUpperIndex &&
+           catalog[pairs[liberalLowerIndex * 2]].spatial *
+                   catalog[pairs[liberalLowerIndex * 2 + 1]].spatial >=
+               maxQueryCos) {
+        liberalLowerIndex++;
+    }
 
     // step the upper index backward
     while (liberalLowerIndex < liberalUpperIndex
-           // the liberalUpperIndex is past the end of the logically returned range, so we need to subtract 1
-           && catalog[pairs[(liberalUpperIndex-1)*2]].spatial * catalog[pairs[(liberalUpperIndex-1)*2+1]].spatial <= minQueryCos
-        )
-    { liberalUpperIndex--; }
+           // the liberalUpperIndex is past the end of the logically returned range, so we need to
+           // subtract 1
+           && catalog[pairs[(liberalUpperIndex - 1) * 2]].spatial *
+                      catalog[pairs[(liberalUpperIndex - 1) * 2 + 1]].spatial <=
+                  minQueryCos) {
+        liberalUpperIndex--;
+    }
 
     *end = &pairs[liberalUpperIndex * 2];
     return &pairs[liberalLowerIndex * 2];
 }
 
 /// Number of star pairs stored in the database
-long PairDistanceKVectorDatabase::NumPairs() const {
-    return index.NumValues();
-}
+long PairDistanceKVectorDatabase::NumPairs() const { return index.NumValues(); }
 
-/// Return the distances from the given star to each star it's paired with in the database (for debugging).
-std::vector<float> PairDistanceKVectorDatabase::StarDistances(int16_t star, const Catalog &catalog) const {
+/// Return the distances from the given star to each star it's paired with in the database (for
+/// debugging).
+std::vector<float> PairDistanceKVectorDatabase::StarDistances(int16_t star,
+                                                              const Catalog &catalog) const {
     std::vector<float> result;
     for (int i = 0; i < NumPairs(); i++) {
-        if (pairs[i*2] == star || pairs[i*2+1] == star) {
-            result.push_back(AngleUnit(catalog[pairs[i*2]].spatial, catalog[pairs[i*2+1]].spatial));
+        if (pairs[i * 2] == star || pairs[i * 2 + 1] == star) {
+            result.push_back(
+                AngleUnit(catalog[pairs[i * 2]].spatial, catalog[pairs[i * 2 + 1]].spatial));
         }
     }
     return result;
 }
 
+///////////////////// Tetra database //////////////////////
+
+long SerializeTetraDatabase(const Catalog &catalog, float maxFovDeg, unsigned char *buffer,
+                            const std::vector<short> &pattStars,
+                            const std::vector<short> &catIndices, bool ser) {
```

Comment:
While the current convention of having a separate function for serialize and serializelength is pretty stupid, until we replace it with serialize functions that return a vector (see #63), we should stick to that and have two separate functions (which could both call a helper function that does the main work). Or you could just fix #63 by returning a vector instead (and modifying other serializers to also return a vector).

Reply:
I'll definitely fix that issue soon, it's painful to double the generation time for no reason. For this pr I'll just split it up for now